[DX][WebProfilerBundle] Add Pretty Print functionality for Request Content

SamFleming opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no   
| Deprecations? | no 
| Tests pass?   | yes 
| Fixed tickets | 
| License       | MIT
| Doc PR        | n/a ?

## Why?
Quite often when attempting to debug issues with JSON requests sent to a Symfony API, I use the Web Profiler to check the request content. More often than not the request content isn't easily readable (99% of the time it's all stuck on a single line and impossible to read). I always find myself copying + pasting the content into a random online tool to have it "pretty-print" the JSON.

Usually this isn't an issue, but can be annoying when offline. There's also the security issue of sending entire JSON payloads to a third-party server just for formatting 😳. Alternatively, maybe developers  copy+paste into their chosen editors and this PR is all a waste of time — I hope not 😛.

## How?
This PR adds "Pretty-Print" JSON functionality straight into the profiler. 

We can use `collector.requestheaders` to detect if the request was JSON and conditionally show the Pretty Print button.

When the button is clicked, we format the JSON from the "Request Content" card.
 
## What does it look like?
Before:
![without-pretty-print](https://user-images.githubusercontent.com/573318/47180751-36b0ce00-d319-11e8-86ed-eb0d78ebcbe3.png)

After:
![pretty](https://user-images.githubusercontent.com/573318/47180763-3c0e1880-d319-11e8-995d-eba565aad827.png)

Non-JSON Requests (unchanged):
![non-json-request](https://user-images.githubusercontent.com/573318/47181080-03227380-d31a-11e8-8cf2-e8b2e8c1a21d.png)

## Things to consider

- Is `JSON.stringify(JSON.parse(content));` the safest, most efficient way to do this?
- Should the "Pretty Print" button be in-line next to the "Request Content" header? I couldn't find a pattern for this sort of thing elsewhere in the profiler.
- Do people want JSON formatted with 4 spaces, would 2 spaces be preferred? Should this be a configuration option stored in localStorage (such as the light/dark theme configuration)?
- Should this be a toggle? E.g. click to pretty print, then click to undo

## Future Improvements

Depending on how this is received it could be extended to support formatting different request content-types (e.g. XML formatting) — I assume.

## Progress

- [x] Gather feedback and decide where to perform the pretty-print: [server-side, or client-side](https://github.com/symfony/symfony/pull/28919#issuecomment-431508361).
*It was decided server-side would be better.*

javiereguiluz commented:
I like this proposal a lot. I would apply this behaviour by default, without any button or toggle. To prevent issues with complex queries, I'd apply a `max-height` of around 500px to the container element.

ro0NL commented:
I think it should be pretty by default, but a `Raw` option could still be useful?

stof commented:
and if we want to do it all the time (with a tab switcher to view the raw one), it would make sense to do the pretty-printing server-side.

SamFleming commented:
👍 Sounds good. I'll look at getting that implemented. 

If we're doing the pretty-printing on the server, that would mean we're putting the entire payload into the DOM twice, once "pretty-printed", the other raw – I'm assuming most payloads aren't too big so this won't be an issue - are we okay with this?

ro0NL commented:
what about "once" server side: `?panel=request&raw=1`

also instead of pretty JSON, wouldnt VarDumper (thus collapsible) be far more useful?

SamFleming committed:
[DX][WebProfilerBundle] Add Pretty Print functionality for Request Content

SamFleming commented:
Okay, so, I've taken the above feedback, rebased and updated the UI. There are now "Pretty" and "Raw" tabs. I've left the formatting down to JS for reasons discussed later.

![screenshot 2018-10-19 at 22 04 05](https://user-images.githubusercontent.com/573318/47243808-2a954100-d3eb-11e8-94b2-3211fcaad64c.png)
![screenshot 2018-10-19 at 22 04 33](https://user-images.githubusercontent.com/573318/47243811-2a954100-d3eb-11e8-8f0d-c53bca427f9d.png)

I definitely prefer this. Thank you for the suggestions 👍 

> and if we want to do it all the time (with a tab switcher to view the raw one), it would make sense to do the pretty-printing server-side.

I don't have much experience with the internals of the profiler. However, from what I can workout there are a few options we have to pretty print on the server.

I'm assuming the method of pretty-printing we'd use is `json_encode(json_decode($requestContent), JSON_PRETTY_PRINT))` (remember, at the moment our `$requestContent` is a string). Alternatively, we can look for additional libraries, but I'm not sure what the stance on that is?

1. `collector` is basically `$this->data` from [`RequestDataCollector.php`](https://github.com/symfony/symfony/blob/v4.1.6/src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php#L89-L112). We could store an additional key (e.g. `"content_formatted"`) and perform our pretty-printing during collection. 
**Pros:** Don't have to pretty print in the profiler (is that a pro?) 
**Cons:** Performance affect during collection; storing the entire request payload twice in the profiler. 
2. We perform the print-printing in [`ProfilerController::panelAction`](https://github.com/SamFleming/symfony/blob/d56ed851b4f2d012c8633f0632feb96d17b96c17/src/Symfony/Bundle/WebProfilerBundle/Controller/ProfilerController.php#L76) and add a key to the available twig variables https://github.com/SamFleming/symfony/blob/d56ed851b4f2d012c8633f0632feb96d17b96c17/src/Symfony/Bundle/WebProfilerBundle/Controller/ProfilerController.php#L104-L112. 
**Pros:** Not storing duplicate content. 
**Cons:** Placing Collector specific logic in the generic `panelAction` — should it really be the responsibility of `panelAction` to look into the Collector and its attributes to perform the printing?
3. Something obvious I'm missing? 😝 🤞 

It would be great to hear some feedback on the above options as I'm not really sure which route to take. 

> also instead of pretty JSON, wouldnt VarDumper (thus collapsible) be far more useful?

I had been thinking about this. My main motivation for not using VarDumper was maintaining the ability to copy the pretty-printed output and edit it (e.g. for use in [Postman](https://www.getpostman.com/)). If we output using the `VarDumper` copy + paste wouldn't be possible. Perhaps we could have an additional "Dump" tab which contains the output — then we have three variations of request content though 🤔 

---

**Sidenote:**

I've had to create commit d56ed851b4f2d012c8633f0632feb96d17b96c17 as there appears to be a bug when nesting tabs in the profiler panels. Looks like until this PR it's never been required. Shall I split this out into a separate PR?

stof commented:
You could add a getter on the Collector object returning the pretty-printed JSON, without the need to do it at collect time (it can be computed from the raw content at render time)

dunglas commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/request.html.twig:
```diff
@@ -171,15 +171,35 @@
-                    <div class="card">
-                        <pre class="break-long-words">{{ collector.content }}</pre>
+                    <div class="sf-tabs">
+                        {% if collector.requestheaders.get('content-type')|lower is same as ('application/json') %}
```
Can you use a regex instead to also prettify `application/ld+json`, `application/hal+json` and so on?

SamFleming replied:
Good catch. I've used `application/([a-z]+\+)?json`, however, am open to suggestions on improving this.

dunglas commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/request.html.twig:
```diff
@@ -171,15 +171,35 @@
                 {{ include('@WebProfiler/Profiler/bag.html.twig', { bag: collector.requestheaders, labels: ['Header', 'Value'], maxDepth: 1 }, with_context = false) }}
 
-                <h3>Request Content</h3>
+                <h3>
```
Should be reverted to prevent merge conflicts

SamFleming replied:
👌

ro0NL commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -552,7 +552,7 @@
                 /* create the tab navigation for each group of tabs */
                 for (var i = 0; i < tabGroups.length; i++) {
-                    var tabs = tabGroups[i].querySelectorAll('.tab');
+                    var tabs = tabGroups[i].querySelectorAll(':scope > .tab');
```
see the top comment of this file. Also; why is this need practically? Doesnt `tabGroups[i]` already defines the scope :S

ro0NL replied:
Got ya. Missed the comment sorry :)

>  however, I didn't want to edit more than I had to for this to work

same features across files should be kept in sync. 

Good to know here is; if the exception page occurs _outside_ the profiler (which has tabs) AND the profiler toolbar is active, then base_js from the profiler applies. Without the toolbar base_js from twig bundle applies.

ro0NL commented on src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php:
```diff
@@ -263,11 +263,26 @@ public function getContent()
         return $this->data['content_type'];
     }
 
+    public function getRequestContentType()
```
public method is unused, perhaps better to inline into `isRequestJSON` until proven needed

SamFleming replied:
I assume the fact I needed it would have proved it needed, but happy to inline if you think that's best.

ro0NL commented on src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php:
```diff
@@ -263,11 +263,26 @@ public function getContent()
+        return $this->data['request_headers']['content-type'];
+    }
+
+    public function isRequestJSON()
```
`isJsonRequest()`

ro0NL commented on src/Symfony/Component/HttpKernel/Tests/DataCollector/RequestDataCollectorTest.php:
```diff
@@ -333,4 +333,32 @@ private function getCookieByName(Response $response, $name)
+    /**
+     * @dataProvider provideJSONTypes
+     */
+    public function testIsJSON($contentType, $expected)
```
`Json`

ro0NL commented on src/Symfony/Component/HttpKernel/Tests/DataCollector/RequestDataCollectorTest.php:
```diff
@@ -333,4 +333,32 @@ private function getCookieByName(Response $response, $name)
+        $this->assertSame($expected, $c->isRequestJSON());
+    }
+
+    public function provideJSONTypes()
```
`provideJsonContentTypes`

ro0NL commented on src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php:
```diff
@@ -263,11 +263,26 @@ public function getContent()
         return $this->data['content'];
     }
 
+    public function getPrettyJSONContent()
```
`Json`

ro0NL commented on src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php:
```diff
@@ -263,11 +263,26 @@ public function getContent()
+
+    public function isRequestJSON()
+    {
+        return 1 === preg_match('{^application/[a-z]*?\+?json$}i', $this->getRequestContentType());
```
matches `application/foojson` .. not sure it's intended. If not: `^application/(?:[a-z]+\+)?json$` is maybe better

ro0NL commented:
We could also add a 3rd tab: "Debug" that uses `profiler_dump()` whereas "Pretty" uses `json_encode` (both available in twig)

ro0NL reviewed (approved):
depending on https://github.com/symfony/symfony/pull/28919#discussion_r236036111

SamFleming commented:
Thanks for the approval @ro0NL – I've updated the regex as suggested 👌 — It will match `application/hal+ld+json`, are you okay with that – alternatively we could change the quantifier to `?` to only match one or zero (instead of zero or more) – your thoughts here would be appreciated. 

I'm currently trying to setup the "Debug" tab as suggested but am struggling with some unexpected errors. I'll keep trying.

ro0NL commented:
@SamFleming perhaps better to introduce the debug tab in a new PR so we dont block/postpone this one (i was just brainfarting :)) But i definitively think it's a nice addition.

> It will match application/hal+ld+json, are you okay with that

yes, i tweaked it to more generic. Now matches `foo+bar+json, foo+++json`. I think we want to be tolerant yes.

SamFleming commented:
> @SamFleming perhaps better to introduce the debug tab in a new PR so we dont block/postpone this one (i was just brainfarting :)) But i definitively think it's a nice addition.

Awesome, 👌 

Anything left for me to do to get this ready for merging?

ro0NL commented:
let's wait :) i do have one last question though; what happens if the request content is not valid JSON? Even if the content-type says so.. does it crash the panel? Will developer see the error?

SamFleming commented:
![image](https://user-images.githubusercontent.com/573318/49035754-9ec2c180-f1ad-11e8-85c3-cca2006adb1b.png)

Is what shows when the JSON is invalid. `json_decode` will return `NULL` when an invalid JSON is encountered, then the `json_encode(NULL)` returns `null`.

We could check `json_last_error` then display `json_last_error_msg` in the "Pretty" tab if there's an issue?

ro0NL commented:
Perhaps fallback to raw content only without tabs then. I.e. the current behavior

SamFleming commented:
Done https://github.com/symfony/symfony/pull/28919/commits/63c02194e7b63052bf990e484bb504d13c992db4

Is this too... horrible?

ro0NL commented on src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php:
```diff
@@ -263,11 +263,21 @@ public function getContent()
         return $this->data['content'];
     }
 
+    public function getPrettyJsonContent()
```
i think we need to use json_last_error() indeed, as "null" can be a valid JSON value too: https://3v4l.org/I7DM7

i propose the return a structure like `return array('content' => json_encode(...), 'valid' => true/false);` and rename the method to `getPrettyJson()`

ro0NL reviewed (approved):
the !="null" check is implicit, but if the null is valid raw&pretty show the same content anyway, so it doesnt really hurt and avoids 2 tabs showing null :+1:

ro0NL reviewed (approved):
Explicit also fine for me :sweat_smile: t:

SamFleming commented:
Oh... I thought I saw a comment about checking `json_last_error`. I've implemented here https://github.com/symfony/symfony/pull/28919/commits/9446deb10e8ae704b9e93034d67633834aa6aac8

I definitely prefer, as it removes any abiquity. However, if we're going this route, is there any point checking the Content-Type for being JSON? Is it worth attempting the `json_encode(json_decode())` no matter what is in the body, then using `json_last_error` to determine whether the tab shows?

ro0NL commented:
the extra overhead might bite us as we try to parse each request content, every time. Tend to prefer the explicit check here.

SamFleming commented:
I also wonder if it's worth caching the result of  [`getPrettyJson`](https://github.com/symfony/symfony/pull/28919/files#diff-e8f5b14fbfbbeac60fc9f3abe310c3b0R266) seeing as it is being called twice in twig and I'm assume `json_decode(json_encode())` could be quite resource intensive... or assigning it to a variable in twig.

SamFleming commented:
> the extra overhead might bite us as we try to parse each request content, every time. Tend to prefer the explicit check here.

👌 Fair

ro0NL commented:
actually we dont parse during the request :D (im tired), only during display in the profiler. But yeah, i think we should be conservative, and expand as needed.

ro0NL commented:
> I also wonder if it's worth caching the result of [`getPrettyJson`](https://github.com/symfony/symfony/pull/28919/files#diff-e8f5b14fbfbbeac60fc9f3abe310c3b0R266) seeing as it is being called twice in twig and I'm assume `json_decode(json_encode())` could be quite resource intensive... or assigning it to a variable in twig.

correct. We should return string|null (checking for last_error intermediary)

SamFleming commented:
> We should return string|null (checking for last_error intermediary)

Appologies, not quite sure what you mean here. I was thinking this https://github.com/symfony/symfony/pull/28919/commits/91a2b392e99879c19faec020f93aeeb3341fdbb7

ro0NL commented:
return `"string"` for valid JSON, or `null` otherwise. Then in twig `prettyJson is not null` :)

ro0NL reviewed (approved):
:ok_hand:

SamFleming commented:
@dunglas was there anything else you wanted me to do on this?

fabpot commented:
Thank you @SamFleming.
